### ISSUE TITLE
todo!() and format!() produce phantom type-mismatch diagnostics; "Expand macro recursively" returns "Not available"

### ISSUE THREAD

**Issue by fullwoodenshovel** [rust-lang/rust-analyzer#22912]
rust-analyzer version: 0.3.2981-standalone
rustc version: rustc 1.92.0 (ded5c06cf 2025-12-08)
editor or extension: VSCode (apt/deb install, not snap), rust-analyzer extension, auto-update enabled
relevant settings: none non-default related to macros/proc-macro; rust-src component confirmed installed via `rustup component list --installed`

## Description

`todo!()` and `format!()` are reported by rust-analyzer as producing phantom type errors, even though the code compiles and runs fine with `cargo check` / `cargo build` / `cargo run`. This is IDE-only, not a real compiler error.

## Minimal reproduction

```rust
use std::println;

fn foo() -> f32 {
    todo!()
}

fn bar() -> String {
    let thing: i32 = 3;
    format!("{thing}")
}

fn main() {
    foo();
    println!("{}", bar());
}
```

## Expected behavior

- `todo!()` should infer as the never-type `!`, which coerces to `f32`, so `foo` should have no diagnostics.
- `format!("{thing}")` should infer as `String`, so `bar` should have no diagnostics.
- `cargo check` confirms both are correct — no errors from rustc.

## Actual behavior

rust-analyzer reports:
- `expected f32, found ()` (E0308) on the closing `}` of `foo`
- `expected String, found ()` (E0308) on the closing `}` of `bar`

Running **"rust-analyzer: Expand macro recursively"** with the cursor on either `todo!()` or `format!()` returns **"Not available"** rather than expanding.

Note that trying:
```rust
fn main() {
    let a = format!("");
}
```
and hovering over `a` displays `{unknown}` as opposed to `()` or `String`.

## What I've ruled out

- Reproduces in a brand-new, minimal `cargo new` project, not workspace/project-specific.
- Reproduces after a full clean reinstall of the extension and a fresh `~/.config/Code` (renamed old one out of the way, started from nothing).
- `rust-src` component confirmed installed and readable (`ls` on the relevant `core`/`std` source files succeeds).
- Not proc-macro-srv related — no external proc-macro crates involved, and this reproduces even with a correctly-configured `rust-analyzer.procMacro.server` path.
- Originally suspected the VSCode install being a snap package (confinement issues), but the bug persists identically after fully migrating to the official apt/deb VSCode install.

## Possibly related

This may be the same underlying issue as #19873 / #19931 ("`Semantics::expand_macro_call` still returns `None` on some macro calls"), though those focus on nested/impl-block macro calls (e.g. a macro expanding to `format!` internally, or macros in impl blocks), whereas this reproduces with plain, directly-written `todo!()` and `format!()` calls with no nesting or custom macros involved; it may be a distinct (or wider-reaching) instance of the same root cause.

**Comment by ejmount** [comment]
I had a similar setup as above (VScode, albeit on Windows 10, Rust 1.92, RA version `0.3.2981-standalone`) and my current project is thrown off because RA has been interpreting `vec![]` as returning `()`. I suspect the underlying problem is the same. Particularly, the following flags a type mismatch, `expected Vec<i32, Global>, found ()` on the closing brace:
```rust
fn baz() -> Vec<i32> {
    vec![1, 2, 3]
}
```
This seems to be a problem with macros rather than anything to do with type inference because using `Vec::from_iter` instead of `vec![]` is not flagged. 

In addition, the above snippet is flagged even in a fresh project, which compiles successfully if cargo is explicitly called. The same fresh project also reproduces the spurious flags in OP's examples. 

I tried upgrading the rust toolchain to 1.97.1 as I was investigating to write this post and that appears to have fixed the problem.

**Comment by fullwoodenshovel** [comment]
Upgrading to 1.97.1 fixed it for me too. Now types are being inferred correctly, macros are expanding correctly, and hoving over variables display the correct result.

**Comment by ChayimFriedman2** [comment]
When will people learn to search issues 😢
